### PR TITLE
Optimize WinHTTP notification callbacks

### LINKED ISSUE #5036: Improve WinHttpHandler use of notification callbacks
WinHttpHandler currently uses the WINHTTP_CALLBACK_FLAG_ALL_NOTIFICATIONS when setting up the status callback function. This isn't strictly necessary. However, some of the notification flags are required such as WINHTTP_CALLBACK_FLAG_REDIRECT, WINHTTP_CALLBACK_FLAG_READ_COMPLETE and others.

This issue will track optimizing the set of notification required as well as providing a way to turn them all on for debugging/logging purposes.

### PR DESCRIPTION
**Pull request by davidsh**
Modify WinHttpHandler and WinHttpWebSocket notification callbacks to only receive those notifications that are actually being used in the status callback. This will improve performance.

Fixes #5036

### CODE REVIEW

**Comment by davidsh** [comment]
@CIPop @stephentoub PTAL

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 1]
```diff
@@ -1276,13 +1276,16 @@ private void CheckDisposedOrStarted()
             SafeWinHttpHandle requestHandle,
             Interop.WinHttp.WINHTTP_STATUS_CALLBACK callback)
         {
-            // TODO: Issue #5036. Having the status callback use WINHTTP_CALLBACK_FLAG_ALL_NOTIFICATIONS
-            // isn't strictly necessary. However, some of the notification flags are required.
-            // This will be addressed this as part of WinHttpHandler performance improvements.
+            uint notificationFlags =
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: const?

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs [hunk 2]
```diff
@@ -137,10 +137,16 @@ public override string SubProtocol
                         WinHttpException.ThrowExceptionUsingLastError();
                     }
 
+                    uint notificationFlags =
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs]: const?

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs]: fixed

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: fixed

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by stephentoub** [comment]
Small nit in that the consts should begin with a capital, but no need to block on that.  LGTM.

**Comment by davidsh** [comment]
>Small nit in that the consts should begin with a capital, but no need to block on that

ok. I will make a note to fix that in a later PR.

**Pull request merged**